### PR TITLE
fix(ssr): use different language tag when with duplicate langs

### PR DESCRIPTION
**Pull request by yin1999**
## Summary

See the result of [hreflang tag testing tool](https://technicalseo.com/seo-tools/hreflang/) on https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Array:

image

### Problem

1. There are duplicate language tags when build zh-CN page and zh-TW is avaliable.
2. The hreflang tag that used is `lang-region` is not prefered now, we may switch to `lang-script` format now. (as a part of mdn/content#23596)

### Solution

To resolve this, we should also include zh-CN locale when try to resolve language tag for zh-TW.

---

### Before

See problem description.

### After

See testing.

---

## How did you test this change?

run `yarn test:testing`

### CODE REVIEW

File: testing/tests/index.test.ts [hunk 1]
```diff
@@ -283,6 +285,35 @@ test("French translation using English front-matter bits", () => {
   expect(bcd.value.query).toBe("javascript.builtins.Array.toLocaleString");
 });
 
+test("content built zh-CN page with en-US fallback image", () => {
```
Comment (yin1999) [inline testing/tests/index.test.ts:288]:
```suggestion
test("content built zh-CN page for hreflang tag testing", () => {
```

**Comment by yin1999** [comment]
Marked as draft to fix https://github.com/mdn/content/issues/23596#issuecomment-1381518729

File: ssr/render.ts [hunk 2]
```diff
@@ -14,7 +14,15 @@ const PREFERRED_LOCALE = {
...
 };
 
-function htmlEscape(s) {
+// We should use the language tag (e.g. "zh-Hans") instead of the locale.
+// This is a map of locale => language tag.
+// See https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry
+const LANGUAGE_TAGS = new Map([
+  ["zh-CN", "zh-Hans"],
+  ["zh-TW", "zh-Hant"],
+]);
```
Comment (caugner) [inline ssr/render.ts:23]: Can we use a simple frozen object instead, and add the region?

```suggestion
const LANGUAGE_TAGS = Object.freeze({
  'zh-CN': 'zh-Hans-CN',
  'zh-TW': 'zh-Hant-TW',
});
```

File: ssr/render.ts [hunk 3]
```diff
@@ -35,17 +43,17 @@ function getHrefLang(locale, otherLocales) {
...
   const preferred = PREFERRED_LOCALE[hreflang];
   if (preferred) {
     // e.g. `preferred===zh-CN` if hreflang was `zh`
     if (locale !== preferred) {
       // e.g. `locale===zh-TW`
-      if (otherLocales.includes(preferred)) {
+      if (allLocales.includes(preferred)) {
         // If the more preferred one was there, use the locale + region format.
-        return locale;
+        return LANGUAGE_TAGS.get(locale) || locale;
```
Comment (caugner) [inline ssr/render.ts:56]:
```suggestion
        return LANGUAGE_TAGS[locale] ?? locale;
```

**Review by caugner — CHANGES_REQUESTED** [review]
LGTM, just one nit.

File: ssr/render.ts (hunk 2 above)
Comment (yin1999) [reply ssr/render.ts]: I'm not sure should we just use `zh-Hans` for Simplified Chinese (zh-CN) and `zh-Hant` for traditional Chinese (zh-TW) without the region code (`CN` and `TW`). As we are using the shortest language tags that won't be duplicated.

Comment (caugner) [reply ssr/render.ts]: You're right, using `zh-Hans` and `zh-Hant` would be consistent with using `pt` for `pt-BR`.

File: ssr/render.ts [hunk 4]
```diff
@@ -16,7 +16,15 @@ const PREFERRED_LOCALE = {
   zh: "zh-CN",
 };
 
-function htmlEscape(s) {
+// We should use the language tag (e.g. "zh-Hans") instead of the locale.
+// This is a map of locale => language tag.
+// See https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry
+const LANGUAGE_TAGS = Object.freeze({
+  "zh-CN": "zh-Hans-CN",
+  "zh-TW": "zh-Hant-TW",
```
Comment (caugner) [inline ssr/render.ts:24]:
```suggestion
  "zh-CN": "zh-Hans",
  "zh-TW": "zh-Hant",
```

File: testing/tests/index.test.ts [hunk 5]
```diff
@@ -210,6 +212,37 @@ test("French translation using English front-matter bits", () => {
...
+  const htmlFile = path.join(builtFolder, "index.html");
+  const html = fs.readFileSync(htmlFile, "utf-8");
+  const $ = cheerio.load(html);
+  // The built page should not have duplicate hreflang tags,
+  // when zh-TW translation is also available.
+  expect($('link[rel="alternate"]')).toHaveLength(4);
+  expect($('link[rel="alternate"][hreflang="en"]')).toHaveLength(1);
+  expect($('link[rel="alternate"][hreflang="fr"]')).toHaveLength(1);
+  expect($('link[rel="alternate"][hreflang="zh"]')).toHaveLength(1);
+  expect($('link[rel="alternate"][hreflang="zh-Hant-TW"]')).toHaveLength(1);
```
Comment (caugner) [inline testing/tests/index.test.ts:239]:
```suggestion
  expect($('link[rel="alternate"][hreflang="zh-Hant"]')).toHaveLength(1);
```

File: testing/tests/index.test.ts [hunk 6]
```diff
@@ -175,10 +176,11 @@ test("content built French foo page", () => {
   const htmlFile = path.join(builtFolder, "index.html");
   const html = fs.readFileSync(htmlFile, "utf-8");
   const $ = cheerio.load(html);
-  expect($('link[rel="alternate"]')).toHaveLength(3);
+  expect($('link[rel="alternate"]')).toHaveLength(4);
   expect($('link[rel="alternate"][hreflang="en"]')).toHaveLength(1);
   expect($('link[rel="alternate"][hreflang="fr"]')).toHaveLength(1);
   expect($('link[rel="alternate"][hreflang="zh"]')).toHaveLength(1);
+  expect($('link[rel="alternate"][hreflang="zh-Hant-TW"]')).toHaveLength(1);
```
Comment (caugner) [inline testing/tests/index.test.ts:183]:
```suggestion
  expect($('link[rel="alternate"][hreflang="zh-Hant"]')).toHaveLength(1);
```

File: testing/tests/index.test.ts [hunk 7]
```diff
@@ -126,10 +126,11 @@ test("content built foo page", () => {
   expect($('script[src="/static/js/ga.js"]')).toHaveLength(1);
 
   // Because this en-US page has a French translation
-  expect($('link[rel="alternate"]')).toHaveLength(3);
+  expect($('link[rel="alternate"]')).toHaveLength(4);
   expect($('link[rel="alternate"][hreflang="en"]')).toHaveLength(1);
   expect($('link[rel="alternate"][hreflang="fr"]')).toHaveLength(1);
   expect($('link[rel="alternate"][hreflang="zh"]')).toHaveLength(1);
+  expect($('link[rel="alternate"][hreflang="zh-Hant-TW"]')).toHaveLength(1);
```
Comment (caugner) [inline testing/tests/index.test.ts:133]:
```suggestion
  expect($('link[rel="alternate"][hreflang="zh-Hant"]')).toHaveLength(1);
```

File: testing/tests/index.test.ts [hunk 8]
```diff
@@ -228,10 +261,11 @@ test("content built zh-TW page with en-US fallback image", () => {
   const htmlFile = path.join(builtFolder, "index.html");
   const html = fs.readFileSync(htmlFile, "utf-8");
   const $ = cheerio.load(html);
-  expect($('link[rel="alternate"]')).toHaveLength(3);
+  expect($('link[rel="alternate"]')).toHaveLength(4);
   expect($('link[rel="alternate"][hreflang="en"]')).toHaveLength(1);
   expect($('link[rel="alternate"][hreflang="fr"]')).toHaveLength(1);
   expect($('link[rel="alternate"][hreflang="zh"]')).toHaveLength(1);
+  expect($('link[rel="alternate"][hreflang="zh-Hant-TW"]')).toHaveLength(1);
```
Comment (caugner) [inline testing/tests/index.test.ts:268]:
```suggestion
  expect($('link[rel="alternate"][hreflang="zh-Hant"]')).toHaveLength(1);
```

**Review by caugner — APPROVED** [review]
Tested locally (`yarn dev` and looked at the source of http://localhost:5042/en-US/docs/Learn/HTML), and works.

**Pull request merged**